### PR TITLE
Add a warning log when a module registers to a hook but implements no hook listener

### LINKED ISSUE #19230: Notify when a module registers to a hook but implements no hook listener
Following #15702

#### Is your feature request related to a problem?

When a module registers to a hook using `$this->registerHook()`, no check is performed to make sure that said module implements a hook method to listen to those hooks. This could happen for example when there's a typing mistake in the hook name.

#### Describe the solution you'd like

In debug mode, throw an exception like this when that happens:

> Hook with the name %hookName% has been registered by %moduleClassName%, but the corresponding method %functionName% has not been defined in the Module class.

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add a warning log when a module registers to a hook but implements no hook listener. Thanks @matthieu-rolland for the initial job
| Type?             | new feature
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #19230
| How to test?      | Cf. #19230

## Errors during the install (blocking this PR)
* :heavy_check_mark:  "Hook with the name displayProductAdditionalInfo has been registered by BlockWishlist, but the corresponding method hookDisplayProductAdditionalInfo has not been defined in the Module class."
  * PrestaShop/blockwishlist/pull/163
  * PrestaShop/blockwishlist/pull/143
  * #28339
* :heavy_check_mark:  "Hook with the name registerGDPRConsent has been registered by Psgdpr, but the corresponding method hookRegisterGDPRConsent has not been defined in the Module class."
  * PrestaShop/psgdpr/pull/161
  * PrestaShop/psgdpr/pull/163
  * #27962
* :heavy_check_mark: "Hook with the name actionProductAttributeDelete has been registered by Ps_EmailAlerts, but the corresponding method hookActionProductAttributeDelete has not been defined in the Module class."
  * PrestaShop/ps_emailalerts/pull/106
  * PrestaShop/ps_emailalerts/pull/108
  * #27923

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27683)

### CODE REVIEW

**Review by matthieu-rolland — DISMISSED** [review]

**Review by jolelievre — COMMENTED** [review]
Except for other PHPStan errors it seems ok to me

It also seems like you'll need to fix the modules from test resources if you want integration tests to succeed

File: classes/Hook.php [hunk 1]
```diff
@@ -531,6 +532,22 @@ public static function registerHook($module_instance, $hook_name, $shop_list = n
                 return false;
             }
 
+            // Check that hook listener is implemented by the module
+            if (!Hook::isHookCallableOn($module_instance, $hook_name) && !($module_instance instanceof WidgetInterface)) {
```
Comment (saulaski) [inline classes/Hook.php]:
```suggestion
            if (!static::isHookCallableOn($module_instance, $hook_name) && !($module_instance instanceof WidgetInterface)) {
```

File: classes/Hook.php [hunk 2]
```diff
@@ -531,6 +532,22 @@ public static function registerHook($module_instance, $hook_name, $shop_list = n
...
+                $message = sprintf(
+                    'Hook with the name %s has been registered by %s, but the corresponding method %s has not been defined in the Module class.',
+                    $hook_name,
+                    get_class($module_instance),
+                    sprintf('hook%s', ucfirst($hook_name))
```
Comment (saulaski) [inline classes/Hook.php]:
```suggestion
                    static::getMethodName($hook_name)
```

File: classes/Hook.php [hunk 3]
```diff
@@ -531,6 +532,22 @@ public static function registerHook($module_instance, $hook_name, $shop_list = n
...
+                );
+                /** @phpstan-ignore-next-line */
+                if (_PS_MODE_DEV_) {
+                    throw new PrestaShopException($message);
```
Comment (saulaski) [inline classes/Hook.php]: More specific and appropriate ?
```suggestion
                    throw new PrestaShopModuleException($message);
```

**Review by saulaski — CHANGES_REQUESTED** [review]

**Review by matks — COMMENTED** [review]
PHP CS Fixer wants
```
/* @phpstan-ignore-next-line */
```
instead of
```
/** @phpstan-ignore-next-line */
```

**Review by kpodemski — DISMISSED** [review]
PHP Stan is not happy.

File: classes/Hook.php (hunk 2 above)
Comment (Progi1984) [reply classes/Hook.php]: Fixed

File: classes/Hook.php (hunk 1 above)
Comment (Progi1984) [reply classes/Hook.php]: Fixed

File: classes/Hook.php (hunk 3 above)
Comment (Progi1984) [reply classes/Hook.php]: Fixed

**Review by jolelievre — APPROVED** [review]
Thanks @Progi1984

**Review by matthieu-rolland — APPROVED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @Progi1984 ,

As discussed, here's a recap of the things I tested.

> What should happen : 
> Uninstall one of these 3 modules BlockWishlist, Psgdpr ou Ps_EmailAlerts, then try to reinstall it in any version but the latest one.
> - mode debug ON, you will see an exception
> - mode debug OFF, you will see a Warning Log

In my case, tested on develop branch.
- I uninstalled blockwishlist module v2.1.0
- I refreshed my page
- I installed wishlist module in v2.0.1
- mode debug ON : I have an error ✔️  
Screenshot 2022-05-05 at 16 02 40

- mode debug OFF : the installation of blockwishlist module v2.1.0 works and I don't have any Warning Log only Informative Log :
Screenshot 2022-05-05 at 16 06 41
Screenshot 2022-05-05 at 16 18 17

**Review by florine2623 — APPROVED** [review]
Thanks for the PR @Progi1984 ,

Here are the different steps I followed.

**Tested with [GDPR v1.4.0](https://github.com/PrestaShop/psgdpr/releases/tag/v1.4.0) :**
- simple uninstall of gdpr module v1.4.1
- refresh page
- reinstall of gdpr module v1.4.0
- error in modal is displayed
Screenshot 2022-05-05 at 17 00 52

- debug mode : NO
- go to module > module manager > upload gdpr module v1.4.0
- install OK
-  configure module : error 500
- Advanced parameter > Logs : no Warning

=> Won't show any warnings because of https://github.com/PrestaShop/psgdpr/pull/161

**Tested with [GDPR v1.2.0](https://github.com/PrestaShop/psgdpr/releases/tag/v1.2.0) :**
- uninstall gdpr module with deletion of module folder
- refresh page
- reinstall of gdpr module v1.2.0 with mode prod
- Advanced parameter > Logs : Warning log is displayed :
Screenshot 2022-05-05 at 17 18 51

**Tested with [ps_emailalerts v2.3.2](https://github.com/PrestaShop/ps_emailalerts/releases/edit/v2.3.2) :**
The warning is displayed correctly with Prod mode :
Screenshot 2022-05-05 at 17 36 39


And error is displayed with debug mode :
Screenshot 2022-05-05 at 17 34 33

**Tested with [blockwishlist v2.0.1](https://github.com/PrestaShop/blockwishlist/releases/tag/v2.0.1) and [blockwishlist v2.0.0](https://github.com/PrestaShop/blockwishlist/releases/tag/v2.0.0) :**
Mode prod :
Screenshot 2022-05-05 at 17 42 53
Mode debug :
Screenshot 2022-05-05 at 17 44 57


It is working as expected !
It is QA ✅

**Comment by Progi1984** [comment]
Thanks @florine2623 & @matthieu-rolland

**Pull request merged**